ocaml/xapi/repository.ml

```diff
@@ -583,3 +702,291 @@ let get_repository_handler (req : Http.Request.t) s _ =
   else
     (error "Rejecting request: local pool repository is not enabled";
      Http_svr.response_forbidden ~req s)
+
+let string_of_updateinfo ui =
+  Printf.sprintf "id=%s rec_guidances=%s abs_guidances=%s guidance_applicabilities=%s"
+    ui.id
+    (String.concat ";" (List.map string_of_guidance ui.rec_guidances))
+    (String.concat ";" (List.map string_of_guidance ui.abs_guidances))
+    (String.concat ";" (List.map string_of_applicability ui.guidance_applicabilities))
+
+let parse_guidances gs =
+  List.map (fun g ->
+      match g with
+      | Xml.Element ("guidance", _, [Xml.PCData v]) ->
+        guidance_of_string v
+      | _ ->
+        error "Unknown node in <absolute|recommended_guidances>";
+        raise Api_errors.(Server_error (invalid_updateinfo_xml, []))
+  ) gs
+
+let parse_applicabilities apps =
```

Comment:
It's a shame these bits of information aren't transported using json, because then we could use a library to parse it (https://github.com/ahrefs/atd) - I'm not aware of an equivalent library for xml parsing

(I assume there is a good reason for using xml?)

Reply:
This xml is defined and used by YUM utilities (and DNF in future).